internal/messagequeue/messagequeue.go

```diff
@@ -539,74 +545,74 @@ func (mq *MessageQueue) pendingWorkCount() int {
 
 // Convert the lists of wants into a Bitswap message
 func (mq *MessageQueue) extractOutgoingMessage(supportsHave bool) bsmsg.BitSwapMessage {
-	mq.wllock.Lock()
-	defer mq.wllock.Unlock()
-
 	// Get broadcast and regular wantlist entries.
-	// SortedEntries() slows down the MessageQueue a lot, and entries only need
-	// to be sorted if the number of wants will overflow the size of the
-	// message (to make sure that the highest priority wants are sent in the
-	// first message).
-	// We prioritize cancels, then regular wants, then broadcast wants.
-	var peerEntries []bswl.Entry
-	var bcstEntries []bswl.Entry
-	maxCancelsSize := mq.cancels.Len() * bsmsg.MaxEntrySize
-	maxPeerSize := mq.peerWants.pending.Len() * bsmsg.MaxEntrySize
-	maxBcstSize := mq.bcstWants.pending.Len() * bsmsg.MaxEntrySize
-
-	if maxCancelsSize+maxPeerSize < mq.maxMessageSize {
-		peerEntries = mq.peerWants.pending.Entries()
-	} else {
-		peerEntries = mq.peerWants.pending.SortedEntries()
-	}
-	if maxCancelsSize+maxPeerSize+maxBcstSize < mq.maxMessageSize {
-		bcstEntries = mq.bcstWants.pending.Entries()
-	} else {
-		bcstEntries = mq.bcstWants.pending.SortedEntries()
+	mq.wllock.Lock()
+	peerEntries := mq.peerWants.pending.Entries()
+	bcstEntries := mq.bcstWants.pending.Entries()
+	cancels := mq.cancels.Keys()
+	if !supportsHave {
+		filteredPeerEntries := peerEntries[:0]
+		// If the remote peer doesn't support HAVE / DONT_HAVE messages,
+		// don't send want-haves (only send want-blocks)
+		//
+		// Doing this here under the lock makes everything else in this
+		// function simpler.
+		//
+		// TODO: We should _try_ to avoid recording these in the first
+		// place if possible.
+		for _, e := range peerEntries {
+			if e.WantType == pb.Message_Wantlist_Have {
+				mq.peerWants.RemoveType(e.Cid, pb.Message_Wantlist_Have)
+			} else {
+				filteredPeerEntries = append(filteredPeerEntries, e)
+			}
+		}
+		peerEntries = filteredPeerEntries
 	}
+	mq.wllock.Unlock()
 
-	// Size of the message so far
-	msgSize := 0
+	// We prioritize cancels, then regular wants, then broadcast wants.
 
-	// Always prioritize cancels, then targeted, then broadcast.
+	var (
+		msgSize         = 0 // size of message so far
+		sentCancels     = 0 // number of cancels in message
+		sentPeerEntries = 0 // number of peer entries in message
+		sentBcstEntries = 0 // number of broadcast entries in message
+	)
 
 	// Add each cancel to the message
-	cancels := mq.cancels.Keys()
 	for _, c := range cancels {
+		msgSize += mq.msg.Cancel(c)
+		sentCancels++
+
 		if msgSize >= mq.maxMessageSize {
-			break
+			goto FINISH
 		}
-		msgSize += mq.msg.Cancel(c)
+	}
 
-		// Clear the cancel - we make a best effort to let peers know about
-		// cancels but won't save them to resend if there's a failure.
-		mq.cancels.Remove(c)
+	// Add each regular want-have / want-block to the message, sorting first
```

Comment:
Let's add a comment explaining why we sort / dont sort

Reply:
Done.